src/compiler/checker.ts

```diff
@@ -12181,6 +12181,20 @@ namespace ts {
                         undefined;
                 }
                 if (t.flags & TypeFlags.Index) {
+                    if ((t as IndexType).type.flags & TypeFlags.IndexedAccess) {
+                        const indexedAccess = (t as IndexType).type as IndexedAccessType;
+                        const baseObjectType = getBaseConstraint(indexedAccess.objectType);
+                        const baseIndexType = getBaseConstraint(indexedAccess.indexType);
+                        const indexedAccessType = baseObjectType && baseIndexType && getIndexedAccessTypeOrUndefined(baseObjectType, baseIndexType);
+                        const mappedIndexTypeOfIndexedAccess = indexedAccessType && mapType(indexedAccessType, getIndexType);
+                        const narrowed = mappedIndexTypeOfIndexedAccess && mapType(keyofConstraintType, t => {

```

Comment:
```suggestion
                        const intersected = mappedIndexTypeOfIndexedAccess && getIntersectionType(keyofConstraintType, mappedIndexTypeOfIndexedAccess);
                        const narrowed = intersected.flags & TypeFlags.Never ? undefined : intersected;
```
should be equivalent and much simpler.

Reply:
IIRC I intentionally haven't written it like that - although rn I don't exactly remember **why**. 

Those are not equivalent as my version filters each member of the `keyofConstraintType` individually, so I end up with a subset of `keyofConstraintType`. Yours simply intersects `keyofConstraintType` with `mappedIndexTypeOfIndexedAccess` and that will **preserve** subset of the `mappedIndexTypeOfIndexedAccess`. So the return types of those operations are not the same.

As I said - right now I don't recall right now why I've chosen this approach but I remember that it was on purpose. That, of course, doesn't mean that my version is better here - perhaps yours is. 

I feel like there is probably a test case to be written here that would help us to determine which one is correct...

Should this be assignable?
```ts
export type Context = {
    V1: { "a": string },
    V2: { "b": string },
}

export function path<V extends keyof Context, K extends keyof Context[V]>(v: V, k: K)  { 
    // should this be assignable?
    const a: 'a' | 'b' = k
    return a
}
```

With my version of the code it isn't but with yours it is.

Funny enough this indexed access errors with both versions:
```ts
export type Context = {
    V1: { "a": string },
    V2: { "b": string },
}

declare const c: {
    a: string;
    b: string
}

export function path<V extends keyof Context, K extends keyof Context[V]>(k: K)  { 
    const a: 'a' | 'b' = k
    return c[k] // this errors
}
```

So it probably should be fixed too if your proposed change is correct.

Reply:
Actually, I think that your version wouldn't be correct - mainly because of the presented case with this being assignable:
```ts
const a: 'a' | 'b' = k
```

We can't guarantee this. After all, as `K` might only be instantiated with `'a'`, `'b` or `never`. Those are the only (?) valid calls here:
```ts
path('V1', 'a')
path('V2', 'b')

declare const v: keyof Context
path(v, {} as never)
```

`keyof` is not distributable so `keyof ({ a: string } | { b: string })` is simply never as it gets the common set of properties.

Reply:
`keyofConstraintType` is `string | number | symbol`, afaik if `mappedIndexTypeofIndexedAccess` is `T`, the operation here is essentially `(string & T) | (number & T) | (symbol & T)`, which is fundamentally what `T & (string | number | symbol)` works out to, no? Unless I'm missing something very subtle?

Reply:
My operation is closer to this:
```ts
type Process<T1, T2> = T1 extends any
  ? [T1 & T2] extends [never]
    ? never
    : T1
  : never;

type FilteredSource = Process<PropertyKey, "a" | "b">;
//   ^? string

type Intersected = PropertyKey & ("a" | "b");
//   ^? 'a' | 'b
```

[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBACgTgewMYQM6oDwBUCMAaKLAJgD4oBeQnKCAD2AgDsATVKAQ0ZACgooB+KAG1cUAGSEiAXRr0mrYYwgA3CHCm8+AqEtVxNfAFxVNx3WoDc3bqEhQAYgEsANgzgRmAZQQBXOCgpYRBR0DHgESDhQAGkIEAIAInYEqAAfKASAIwSSKwB6PK0APX5rW2gASUY3VAgkBmZA8MiYuPEoAAoklPSshIBKfMK+Eu4gA)

Reply:
Ah, I missed that it was returning `t` and not the intersection result in the non-never case.

You wanna use `filterType` instead of `mapType` then.

Reply:
> You wanna use filterType instead of mapType then.

I've pushed out this change.

q: do you maybe have an idea for a better test than the first snippet [here](https://github.com/microsoft/TypeScript/pull/49107#discussion_r880334006)? or do you think that trying to write a "reversed" test here is not worth it/bad idea?